pkg/operations/kubernetesupgrade/upgrader.go

```diff
@@ -216,51 +213,81 @@ func (ku *Upgrader) upgradeAgentPools() error {
 		upgradeAgentNode.Client = ku.Client
 		upgradeAgentNode.kubeConfig = ku.kubeConfig
 
-		upgradedAgentsIndex := make(map[int]bool)
+		upgradeVMs := make(map[int]*UpgradeVM)
 
 		for _, vm := range *agentPool.UpgradedAgentVMs {
 			ku.logger.Infof("Agent VM: %s, pool name: %s on expected orchestrator version\n", *vm.Name, *agentPool.Name)
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
-			upgradedAgentsIndex[agentIndex] = true
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, true}
 		}
 
 		ku.logger.Infof("Starting upgrade of agent nodes in pool identifier: %s, name: %s...\n",
 			*agentPool.Identifier, *agentPool.Name)
 
 		for _, vm := range *agentPool.AgentVMs {
-			ku.logger.Infof("Upgrading Agent VM: %s, pool name: %s\n", *vm.Name, *agentPool.Name)
-
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, false}
+		}
 
-			err := upgradeAgentNode.DeleteNode(vm.Name)
-			if err != nil {
-				ku.logger.Infof("Error deleting agent VM: %s\n", *vm.Name)
-				return err
-			}
+		// Create an auxiliary node if hasn't been created yet.
+		// This node is used to take on the load from deleting nodes
+		addedNode := false
+		if agentCount > 0 && len(upgradeVMs) != agentCount {
+			addedNode = true
+			agentIndex := getAvailableIndex(upgradeVMs)
+			ku.logger.Infof("Adding auxiliary agent node with index %d", agentIndex)
 
 			err = upgradeAgentNode.CreateNode(*agentPool.Name, agentIndex)
 			if err != nil {
-				ku.logger.Infof("Error creating upgraded agent VM: %s\n", *vm.Name)
+				ku.logger.Infof("Error creating auxiliary node\n")
 				return err
 			}
+			upgradeVMs[agentIndex] = &UpgradeVM{"", true}
```

Comment:
huh, don't we want to store it's name here? we depend on it below.  We could make CreateNode return the name of the node created

Reply:
We don't know the name here, and don't need it as we are not going to upgrade this node.
I will add comments